### PR TITLE
Add authentication to openapi Spec

### PR DESCRIPTION
**Pull request by mbohlool**
We need authentication definition in OpenAPI spec to be able to generate a client with authorization.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34691)

### CODE REVIEW

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (caesarxuchao) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: Is "beraer authorization" a kind of "apiKey" authorization?

Comment (mbohlool) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: Yes, because we pass a "key" with name "Authorization" in header. They key is "Bearer $TOKEN"

Comment (mbohlool) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: More: any authentication that passes a key in header or query is called apiKey in OpenAPI. here bearer use "authorization" key in header to pass token.

**Comment by mbohlool** [comment]
@k8s-bot unit test this issue #IGNORE

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (caesarxuchao) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: s/t/to

**Comment by caesarxuchao** [comment]
A typo and a question. Otherwise lgtm. @nikhiljindal could you also take a look? I may miss some details about openapi.

File: api/openapi-spec/api.json
Comment (mbohlool) [inline api/openapi-spec/api.json]: So from the list of authenticators I can see, two can be represented in openapi. Token based and http basic (that is actually using the same header as bearer). I've added both. Some authenticators are reduced to bear token, some are not presentable in openapi.

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (mbohlool) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: Done.

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (nikhiljindal) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: This has the same issue as the one @smarterclayton had pointed out in last PR.
We are adding the openapi dependency everywhere.
Anyone using this authenticator now pulls in the dependency.

Comment (mbohlool) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: I am for less dependencies but I think this is different. Previous one could be resolved by refactoring our code, but for this one, I need to duplicate what I need from openapi spec dependency somewhere (probably openapi/common) and import that. I can do that but not in favor of it because we need to keep two copies in sync.

Comment (nikhiljindal) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: Another alternative is to have a method that takes an auth interface and returns the OpenAPISpec documentation for that. Our apiserver will use that method and hence depend on open api spec. Anyone else who just wants to use the authorizer can still do that without depending on open api spec.

**Comment by smarterclayton** [comment]
Pulling in @kubernetes/sig-auth because they own all the Authenticator interfaces and this may be a breaking change.

Will take a look tomorrow.  We also have chained Authenticators so there is likely to be some complexity about what shows up in the model.  It may be that model selection should be determined where we start the processes rather than deeper in the stack.

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (ericchiang) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: I'm +1 for the `GetOpenAPISecurityDefinition` logic living outside of these packages. `pkg/apiserver/authenticator` seems like a better place since that package understands the name and type (bearer, x509, etc.) of each authenticator.

**Comment by smarterclayton** [comment]
And we don't need to model the complexity of auth in openapi - just enough
for simple clients.

Can openapi represent multiple mechanisms?

On Oct 17, 2016, at 11:39 PM, Eric Chiang <notifications@github.com> wrote:

*@ericchiang* commented on this pull request.
------------------------------

In pkg/auth/authenticator/bearertoken/bearertoken.go
<https://github.com/kubernetes/kubernetes/pull/34691>:

> @@ -20,16 +20,19 @@ import (
 	"net/http"
 	"strings"

+	"github.com/go-openapi/spec"

I'm +1 for the GetOpenAPISecurityDefinition logic living outside of these
packages. pkg/apiserver/authenticator seems like a better place since that
package understands the name and type (bearer, x509, etc.) of each
authenticator.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34691>, or mute the thread

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (liggitt) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: I'd rather it be at the top level in the apiserver config.

**Comment by mbohlool** [comment]
The spec itself allows multiple mechanisms, but clients I've generated currently does not work with multiple mechanism with the same header key (all of our bearer authentication use "authorization" key in the header, so it won't work with them, there is one exception, that is basic authorization and one bearer can work together), but this is a problem of client generator and can be fixed. The spec does allow it, http does allow it (multiple "authorization" values are allowed).

File: pkg/auth/authenticator/bearertoken/bearertoken.go
Comment (mbohlool) [inline pkg/auth/authenticator/bearertoken/bearertoken.go]: @ericchiang pkg/apiserver/authenticator is not the only place that uses authenticators (k8s and federation server.go are at least two other places I see this) so the logic will be fragmented into those places authenticators are used and it will be in any place we use authenticators (now or later). 

I am trying to understand your (also @liggitt) concern here, is it the place of the logic or the change in interfaces or dependencies? I may be able fix the interfaces (no change in interfaces) but keep the logic in authenticators.

@nikhiljindal That is not possible until we depend on switch case on type of authenticators and have access to their internals by adding access methods because of union and wrapper authenticators. If I did not understand what you are suggesting, please explain.

**Comment by mbohlool** [comment]
I've changed the code to remove code from authentications. I should just add the note here that the approach is not my favor. If somebody add any other authenticator, they need to remember to add SecurityDefinition to OpenAPI spec. I am afraid people (and me in 6 month) will forget that.

**Review by liggitt — DISMISSED** [review]

File: cmd/kube-apiserver/app/server.go
Comment (mbohlool) [inline cmd/kube-apiserver/app/server.go]: Done.

File: federation/cmd/federation-apiserver/app/server.go
Comment (mbohlool) [inline federation/cmd/federation-apiserver/app/server.go]: Done.

File: pkg/apiserver/authenticator/authn.go
Comment (mbohlool) [inline pkg/apiserver/authenticator/authn.go]: For this and duplicate bearer definitions, I was thinking this could have a documentation/discovery benefit as users can look at it and see what kind of authentication supported by this server. you are absolutely right that without more information it has less uses and clients are not using these duplicates at all and thread them the same. I would remove them for now.

File: pkg/apiserver/authenticator/authn.go
Comment (mbohlool) [inline pkg/apiserver/authenticator/authn.go]: ditto.

File: pkg/apiserver/authenticator/authn.go
Comment (liggitt) [inline pkg/apiserver/authenticator/authn.go]: a custom object with add methods we call repeatedly on the same auth methods seems overly complicated, when all we really want is to know if we should add a basic security definition and bearer security definition

**Comment by mbohlool** [comment]
PTAL. fixup commit will be squashed into other commits after I got LGTM.

File: pkg/genericapiserver/openapi/common/common.go
Comment (liggitt) [inline pkg/genericapiserver/openapi/common/common.go]: pointer to a map is really weird

File: pkg/apiserver/authenticator/authn.go
Comment (mbohlool) [inline pkg/apiserver/authenticator/authn.go]: Done.

**Comment by liggitt** [comment]
this seems ok. I still question whether openapi will be able to effectively express the authn/authz supported by the API server, but if this at least tells clients they can send basic/bearer auth, I guess that's ok.

**Comment by mbohlool** [comment]
This will only enable clients to send token/basic auth. OpenAPI may not be able to completely express auth of k8s API Server. I will rebase and merge then. Thanks.

**Comment by mbohlool** [comment]
Rebase done. Can one of the reviewers look at this again and apply LGTM if everything is OK please.

**Comment by mbohlool** [comment]
Note: I've removed basic auth from generated spec they we keep in source tree. Those files will be source for client generation and at least one client I've tried (python) had problem with basic and token auth exists together. Username/pass can still pass as the header "authorization" key.

**Comment by caesarxuchao** [comment]
This seems good. I'll apply the lgtm label to enable the following work. We can improve the spec later.

**Comment by mbohlool** [comment]
Updated newly added bazel build files using hack/update-bazel.sh, reapplying lgtm as there were no code change.

**Pull request merged**
